### PR TITLE
[GPU] Ngraph based shape-infer infra

### PR DESCRIPTION
**Pull request by vladimir-paramuzov**

### CODE REVIEW

File: src/plugins/intel_gpu/include/intel_gpu/plugin/graph.hpp [hunk 1]
```diff
@@ -83,14 +83,16 @@ class Graph {
     bool use_external_queue() const;
 
 protected:
+    InferenceEngine::gpu::ClContext::Ptr m_context;
+    std::shared_ptr<Program> m_program;
+    std::string m_networkName;
+    Config m_config;
+    uint16_t m_stream_id;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/plugin/graph.hpp:90]: What is the motivation of this change? will the graph is created per stream???

Comment (yeonbok) [reply src/plugins/intel_gpu/include/intel_gpu/plugin/graph.hpp:90]: Ah.. they were just moved..

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/include/intel_gpu/plugin/graph.hpp:90]: You can look at the last commit only here. Most of the other changes are covered here: https://github.com/openvinotoolkit/openvino/pull/12172

File: src/plugins/intel_gpu/src/graph/one_hot.cpp [hunk 2]
```diff
@@ -27,6 +29,37 @@ static bool is_output_bfzyx(const layout& input, int32_t axis) {
...
+    };
+
+    int64_t depth = desc->depth;
+
+    auto depth_tensor = std::make_shared<ngraph::runtime::HostTensor>(ov::element::i64, ov::Shape{1}, static_cast<void*>(&depth));
+    std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>> const_data = {
+        {1, depth_tensor}
+    };
+    ov::op::v1::shape_infer(&op, input_shapes, output_shapes, const_data);
+    return {{output_shapes[0], dt, format::get_default_format(output_shapes[0].size())}};
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/one_hot.cpp:60]: Also I have been curious how we should deal with different opsets. (e.g., gather_elemetns' output shape is different for the optset). Should we keep the opset in the primitive?

[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/one_hot.cpp:60]: > Are we going to support get_default_format for b or bf in the future in the program graph transformation level? the manipulation or comparison with the 4d default format results in many issues.

I think we'll need to replace tag-based format representation with some memory descriptor (dims order + blocking info  + something else) in the future. 

> Also I have been curious how we should deal with different opsets. (e.g., gather_elemetns' output shape is different for the optset). Should we keep the opset in the primitive?

Not sure if it's really needed, but we can store some additional info to have correct shape infer logic.

File: src/plugins/intel_gpu/src/graph/broadcast.cpp [hunk 3]
```diff
@@ -26,6 +28,31 @@ layout broadcast_inst::calc_output_layout(broadcast_node const& node) {
...
+
+    auto tensor1 = make_host_tensor({ov::PartialShape{4}, data_types::i64, format::bfyx}, target_shape.data());
+    std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>> const_data = {
+        {1, tensor1},
+    };
+    ov::op::v3::shape_infer(&op, input_shapes, output_shapes, const_data);
+    auto output_format = format::get_default_format(output_shapes[0].size());
+
+    return { layout{output_shapes[0], input_layout.data_type, output_format} };
+}
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/broadcast.cpp:54]: BTW, do we need to do shape_infer for broadcast? 
It will be either of the cases 
(1) static : Use the given output_shape set atop creation 
(2) If the output_shape is not given at the op creation 
- Read the input(1)'s output and use it as output shape 
Isnt this enough for this OP?

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/broadcast.cpp:54]: I think the overall rule should be the following:
 - Easy shape infer functions can be implemented directly in the plugin w/o ngraph calls (e.g. for the cases when output_shape = input_shape)
 - Otherwise, ngraph should be called. 
 
Not sure about Broadcast, as it has additional parameter (axes set) which specifies the mapping between input and output dimensions which may require some additional logic for shape inference...

**Review by Lyamin-Roman — APPROVED** [review]

**Review by yeonbok — APPROVED** [review]
